modules/system-disaster-recovery/src/main/java/org/apache/ignite/internal/disaster/system/SystemDisasterRecoveryManagerImpl.java

```diff
@@ -0,0 +1,325 @@
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+import static org.apache.ignite.internal.util.ByteUtils.toBytes;
+import static org.apache.ignite.internal.util.CompletableFutures.allOf;
+import static org.apache.ignite.internal.util.CompletableFutures.nullCompletedFuture;
+import static org.apache.ignite.lang.ErrorGroups.Common.INTERNAL_ERR;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import org.apache.ignite.internal.cluster.management.ClusterState;
+import org.apache.ignite.internal.cluster.management.network.messages.CmgMessagesFactory;
+import org.apache.ignite.internal.cluster.management.network.messages.SuccessResponseMessage;
+import org.apache.ignite.internal.cluster.management.raft.ClusterStateStorage;
+import org.apache.ignite.internal.cluster.management.raft.ClusterStateStorageManager;
+import org.apache.ignite.internal.disaster.system.message.ResetClusterMessage;
+import org.apache.ignite.internal.disaster.system.message.SystemDisasterRecoveryMessageGroup;
+import org.apache.ignite.internal.disaster.system.message.SystemDisasterRecoveryMessagesFactory;
+import org.apache.ignite.internal.lang.ByteArray;
+import org.apache.ignite.internal.lang.IgniteInternalException;
+import org.apache.ignite.internal.manager.ComponentContext;
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.network.MessagingService;
+import org.apache.ignite.internal.network.NetworkMessage;
+import org.apache.ignite.internal.network.TopologyService;
+import org.apache.ignite.internal.util.CompletableFutures;
+import org.apache.ignite.internal.vault.VaultEntry;
+import org.apache.ignite.internal.vault.VaultManager;
+import org.apache.ignite.network.ClusterNode;
+
+/**
+ * Implementation of {@link SystemDisasterRecoveryManager}.
+ */
+public class SystemDisasterRecoveryManagerImpl implements SystemDisasterRecoveryManager, IgniteComponent {
+    private static final String NODE_INITIALIZED_VAULT_KEY = "systemRecovery.nodeInitialized";
+    private static final String CLUSTER_NAME_VAULT_KEY = "systemRecovery.clusterName";
+    private static final String RESET_CLUSTER_MESSAGE_VAULT_KEY = "systemRecovery.resetClusterMessage";
+
+    private final String thisNodeName;
+    private final TopologyService topologyService;
+    private final MessagingService messagingService;
+    private final VaultManager vaultManager;
+    private final ServerRestarter restarter;
+
+    private final ClusterStateStorageManager clusterStateStorageManager;
+
+    private final SystemDisasterRecoveryMessagesFactory messagesFactory = new SystemDisasterRecoveryMessagesFactory();
```

Comment:
The field `messagesFactory` is final. Which one do you refer to?

Statics are evil, it's really hard to get rid of them if we are faced with such necessiry for some reason (which we don't see yet). I would leave these non-static.

Reply:
I'm stupid, I meant `static` =)

Reply:
> if we are faced with such necessiry for some reason

Right now we must suffer because of it. Is it really worth it for an elusive future change that will probably never happen?

Reply:
I don't feel any suffering

Reply:
I made it `static`